lineapy/plugins/airflow_pipeline_writer.py

```diff
@@ -0,0 +1,203 @@
+            ...
+            def task_run_session_including_g2():
+                artifacts = g2_z_module.run_session_including_g2()
+                pickle.dump(artifacts["g2"], open("/tmp/g2_z/artifact_g2.pickle", "wb"))
+                pickle.dump(artifacts["z"], open("/tmp/g2_z/artifact_z.pickle", "wb"))
+            with DAG(...) as dag:
+                run_session_including_g2 = PythonOperator(
+                    task_id="run_session_including_g2_task",
+                    python_callable=task_run_session_including_g2,
+                )
+
+        Example of `PythonOperatorPerArtifact`, if the two artifacts in our pipeline
+        (e.g., model and prediction) were created in the same session:
+        .. code-block:: python
+            import pickle
+            import iris_module
+            ...
+            def task_iris_model():
+                mod = iris_module.get_iris_model()
+                pickle.dump(mod, open("/tmp/iris/variable_mod.pickle", "wb"))
+            def task_iris_pred():
+                mod = pickle.load(open("/tmp/iris/variable_mod.pickle", "rb"))
+                pred = iris_module.get_iris_pred(mod)
+                pickle.dump(
+                    pred, open("/tmp/iris/variable_pred.pickle", "wb")
+                )
+            with DAG(...) as dag:
+                iris_model = PythonOperator(
+                    task_id="iris_model_task",
+                    python_callable=task_iris_model,
+                )
+                iris_pred = PythonOperator(
+                    task_id="iris_pred_task",
+                    python_callable=task_iris_pred,
+                )
+            iris_model >> iris_pred
+
+        This way, the generated Airflow DAG file opens room for engineers
+        to control pipeline runs at a finer level and allows for further customization.
+        """
+        DAG_TEMPLATE = load_plugin_template("airflow_dag_PythonOperator.jinja")
+        if (
+            AirflowDagFlavor[dag_flavor]
+            == AirflowDagFlavor.PythonOperatorPerSession
+        ):
+            task_defs = self.get_session_task_definition()
+        elif (
+            AirflowDagFlavor[dag_flavor]
+            == AirflowDagFlavor.PythonOperatorPerArtifact
+        ):
+            task_defs = self.get_artifact_task_definitions()
+
+        task_params = self.get_task_params_args(task_defs)
+        task_functions = list(task_defs.keys())
+        rendered_task_definitions = self.get_rendered_task_definitions(
+            task_defs
+        )
+        dependencies = {
+            task_functions[i + 1]: {task_functions[i]}
+            for i in range(len(task_functions) - 1)
+        }
+        task_graph = TaskGraph(
+            nodes=task_functions,
+            mapping={f: f for f in task_functions},
+            edges=dependencies,
+        )
+        tasks = [
+            {"name": ft, "op_kwargs": task_params.get(ft, None)}
+            for ft in task_functions
```

Comment:
NIT: Abbreviate to `tf` instead of `ft`?

Reply:
👍